aeon/schema/social.py

```diff
@@ -0,0 +1,94 @@
+"""Readers for data relevant to Social experiments."""
+
+import json
+from pathlib import Path
+
+import pandas as pd
+
+import aeon.io.reader as _reader
+
+
+class Pose(_reader.Harp):
+    """Reader for Harp-binarized tracking data given a model that outputs id, parts, and likelihoods."""
+    def __init__(self, pattern):
+        self.pattern = pattern
+        self.extension = "bin"
+
+    def read(self, file):
+        """
+        Reads data from the Harp-binarized tracking file
+        """
+        # Get config file from `file`, then bodyparts from config file.
+        ceph_proc_dir = "/ceph/aeon/aeon/data/processed"
+        model_dir = file.stem.replace("_", "/")
+        config_file_dir = Path(ceph_proc_dir + "/" + model_dir)
+        assert config_file_dir.exists(), f"Cannot find model dir {config_file_dir}"
+        config_file = get_config_file(config_file_dir)
+        parts = self.get_bodyparts(config_file)
+        if parts is None:
+            raise ValueError(f"Cannot find bodyparts in {config_file}.")
+        # With bodyparts, set columns as set by Bonsai, and read data in default format.
+        columns = ["class", "class_likelihood"]
+        for part in parts:
+            columns.extend([f"{part}_x", f"{part}_y", f"{part}_likelihood"])
+        self.columns = columns
+        data = super().read(file)
+        # Set new columns, and reformat `data`.
+        columns = ["class", "class_likelihood", "part", "part_likelihood", "x", "y"]
+        new_data = pd.DataFrame(columns=columns)
+        for part in parts:
+            part_data = data[["class", "class_likelihood", f"{part}_likelihood", f"{part}_x", f"{part}_y"]]
+            part_data.insert(2, "part", part)
+            part_data.columns = columns
+            new_data = pd.concat([new_data, part_data])
+        return new_data.sort_index()
+       
+    def get_bodyparts(file):
+        """Returns a list of bodyparts from a model's config file."""
+        parts = None
+        if file.stem == "confmap_config":  # SLEAP
+            with open(file) as f:
+                config = json.load(f)
+            try:
+                heads = config["model"]["heads"]
+                for model in heads.keys():
+                    if heads[model] is not None:
+                        parts = heads[model]["confmaps"]["part_names"]
+                        break
+            except KeyError as err:
+               raise KeyError(f"Cannot find bodyparts in {file}.") from err
```

Comment:
Can confirm that "part_names" may exist on either level 2 or level 3 in different models:

- `heads[model]["part_names"]` for "single_instance", "centered_instance"
- `heads[model]["confmaps"]["part_names"]` for models with prefix "multi_"
- not at all for "centroid"
```suggestion
            try:
                    heads = config["model"]["heads"]
                    for model, conf in heads.items():
                        if conf is not None:
                            if "part_names" in conf.keys():
                                parts = conf["part_names"]
                            else:
                                parts = heads[model]["confmaps"]["part_names"]
            except KeyError as err:
               raise KeyError(f"Cannot find bodyparts in {file}.") from err
```